libs/backend/src/scan/cast_vote_records/export.ts

```diff
@@ -0,0 +1,386 @@
+/* istanbul ignore file */
+import { Buffer } from 'buffer';
+import fs from 'fs/promises';
+import { sha256 } from 'js-sha256';
+import path from 'path';
+import { Readable } from 'stream';
+import { assertDefined, err, ok, Result } from '@votingworks/basics';
+import {
+  BallotIdSchema,
+  BallotPageLayout,
+  BatchInfo,
+  CVR,
+  ElectionDefinition,
+  MarkThresholds,
+  PollsState,
+  unsafeParse,
+} from '@votingworks/types';
+import { UsbDrive } from '@votingworks/usb-drive';
+import {
+  BooleanEnvironmentVariableName,
+  generateCastVoteRecordExportDirectoryName,
+  generateElectionBasedSubfolderName,
+  isFeatureFlagEnabled,
+  SCANNER_RESULTS_FOLDER,
+} from '@votingworks/utils';
+
+import { ExportDataError, Exporter } from '../../exporter';
+import { SCAN_ALLOWED_EXPORT_PATTERNS, VX_MACHINE_ID } from '../globals';
+import { buildCastVoteRecord as baseBuildCastVoteRecord } from './build_cast_vote_record';
+import { buildCastVoteRecordReportMetadata as baseBuildCastVoteRecordReportMetadata } from './build_report_metadata';
+import {
+  CanonicalizedSheet,
+  canonicalizeSheet,
+  describeSheetValidationError,
+} from './canonicalize';
+import { ResultSheet } from './legacy_export';
+import { buildElectionOptionPositionMap } from './option_map';
+
+type ExportCastVoteRecordToUsbDriveError =
+  | { type: 'invalid-sheet-found'; message: string }
+  | ExportDataError;
+
+/**
+ * The subset of scanner store methods relevant to exporting cast vote records
+ */
+export interface ScannerStore {
+  clearCastVoteRecordHashes(): void;
+  getBatches(): BatchInfo[];
+  getCastVoteRecordRootHash(): string;
+  getElectionDefinition(): ElectionDefinition | undefined;
+  getExportDirectoryName(): string | undefined;
+  getMarkThresholds(): MarkThresholds;
+  getPollsState(): PollsState | undefined;
+  getTestMode(): boolean;
+  updateCastVoteRecordHashes(
+    castVoteRecordId: string,
+    castVoteRecordHash: string
+  ): void;
+  setExportDirectoryName(exportDirectoryName: string): void;
+}
+
+/**
+ * State that can be retrieved via the ScannerStore interface and that is unchanged by exporting
+ * cast vote records
+ */
+interface ScannerStateUnchangedByExport {
+  batches: BatchInfo[];
+  electionDefinition: ElectionDefinition;
+  inTestMode: boolean;
+  markThresholds: MarkThresholds;
+  pollsState?: PollsState;
+}
+
+/**
+ * A grouping of inputs needed by helpers throughout this file
+ */
+interface ExportContext {
+  exporter: Exporter;
+  scannerStore: ScannerStore;
+  scannerState: ScannerStateUnchangedByExport;
+}
+
+interface File {
+  fileName: string;
+  fileContents: string | Buffer;
+}
+
+/**
+ * Metadata stored in the top-level metadata file for a cast vote record export
+ */
+interface CastVoteRecordExportMetadata {
+  arePollsClosed?: boolean;
+  /** Global data relevant to all cast vote records in an export, e.g. election info */
+  castVoteRecordReportMetadata: CVR.CastVoteRecordReport;
```

Comment:
This is how I'm solving the problem of identifying the "canonical" CVR metadata. Rather than requiring importers to merge and construct the latest metadata from the metadata blobs in individual CVR files, I'm always exporting a latest, up-to-date metadata blob in a top-level metadata file.